cmd/kubeadm/app/cmd/phases/controlplane.go

```diff
@@ -17,186 +17,111 @@ limitations under the License.
-func runCmdControlPlane(cmdFunc func(outDir string, cfg *kubeadmapi.InitConfiguration) error, outDir, cfgPath *string, featureGatesString *string, cfg *kubeadmapiv1beta1.InitConfiguration) func(cmd *cobra.Command, args []string) {
-
-	// the following statement build a closure that wraps a call to a cmdFunc, binding
-	// the function itself with the specific parameters of each sub command.
-	// Please note that specific parameter should be passed as value, while other parameters - passed as reference -
-	// are shared between sub commands and gets access to current value e.g. flags value.
-	return func(cmd *cobra.Command, args []string) {
-		var err error
-		if err = validation.ValidateMixedArguments(cmd.Flags()); err != nil {
-			kubeadmutil.CheckErr(err)
+func runControlPlanePhase(component string) func(c workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(controlPlaneData)
+		if !ok {
+			return errors.New("control-plane phase invoked with an invalid data struct")
 		}
-
-		if cfg.FeatureGates, err = features.NewFeatureGate(&features.InitFeatureGates, *featureGatesString); err != nil {
-			kubeadmutil.CheckErr(err)
+		cfg := data.Cfg()
+
+		// special case to handle audit policy for the API server
+		if component == kubeadmconstants.KubeAPIServer && features.Enabled(cfg.FeatureGates, features.Auditing) {
+			// Setup the AuditPolicy (either it was passed in and exists or it wasn't passed in and generate a default policy)
+			if cfg.AuditPolicyConfiguration.Path != "" {
+				// TODO(chuckha) ensure passed in audit policy is valid so users don't have to find the error in the api server log.
+				if _, err := os.Stat(cfg.AuditPolicyConfiguration.Path); err != nil {
+					return fmt.Errorf("error getting file info for audit policy file %q [%v]", cfg.AuditPolicyConfiguration.Path, err)
+				}
+			} else {
+				cfg.AuditPolicyConfiguration.Path = filepath.Join(data.KubeConfigDir(), kubeadmconstants.AuditPolicyDir, kubeadmconstants.AuditPolicyFile)
+				if err := auditutil.CreateDefaultAuditLogPolicy(cfg.AuditPolicyConfiguration.Path); err != nil {
+					return fmt.Errorf("error creating default audit policy %q [%v]", cfg.AuditPolicyConfiguration.Path, err)
+				}
+			}
 		}
 
-		// This call returns the ready-to-use configuration based on the configuration file that might or might not exist and the default cfg populated by flags
-		internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(*cfgPath, cfg)
-		kubeadmutil.CheckErr(err)
-		err = configutil.VerifyAPIServerBindAddress(internalcfg.APIEndpoint.AdvertiseAddress)
-		kubeadmutil.CheckErr(err)
-
-		if err := features.ValidateVersion(features.InitFeatureGates, internalcfg.FeatureGates, internalcfg.KubernetesVersion); err != nil {
-			kubeadmutil.CheckErr(err)
+		fmt.Printf("[control-plane] creating static Pod manifest for %s", component)
+		if err := controlplane.CreateStaticPodFiles(data.ManifestDir(), cfg, component); err != nil {
```

Comment:
seems to me that the original code in runInit() used the same folder for dry-run and without dry-run in the case of etcd and api server.

here by using ManifestDir() we now respect the dry-run flag.

Reply:
I don't have history here, but what you are proposing seems ok to me.